src/ui/FunctionsOutlineView.cpp

```diff
@@ -0,0 +1,366 @@
+			prefix = "\xe2\x8c\xab";
+			break;
+		case 7:
+			// Class
+			prefix = "\xe2\x8c\xac";
+			break;
+		default:
+			break;
+	}
+
+	if (!prefix.IsEmpty())
+		newText.Append(prefix).Append(" ");
+	newText.Append(text);
+	SetText(newText.String());
+	BStringItem::DrawItem(owner, bounds, complete);
+	SetText(text.String());
+#endif
+}
+
+
+// Sort by line
+static int
+CompareItems(const BListItem* itemA, const BListItem* itemB)
+{
+	const SymbolListItem* A = static_cast<const SymbolListItem*>(itemA);
+	const SymbolListItem* B = static_cast<const SymbolListItem*>(itemB);
+
+	int32 lineA;
+	A->Details().FindInt32("be:line", &lineA);
+
+	int32 lineB;
+	B->Details().FindInt32("be:line", &lineB);
+
+	return lineA - lineB;
+}
+
+
+static int
+CompareItemsText(const BListItem* itemA, const BListItem* itemB)
+{
+	return BPrivate::NaturalCompare(((BStringItem*)itemA)->Text(), ((BStringItem*)itemB)->Text());
+}
+
+
+#undef B_TRANSLATION_CONTEXT
+#define B_TRANSLATION_CONTEXT "FunctionsOutlineView"
+
+
+FunctionsOutlineView::FunctionsOutlineView()
+	:
+	BView(B_TRANSLATE("Class outline"), B_WILL_DRAW),
+	fListView(nullptr),
+	fToolBar(nullptr),
+	fStatus(STATUS_EMPTY),
+	fSymbolsLastUpdateTime(0)
+{
+	SetFlags(Flags() | B_PULSE_NEEDED);
+
+	fListView = new BOutlineListView("listview");
+	fScrollView = new BScrollView("scrollview", fListView,
+		B_FRAME_EVENTS | B_WILL_DRAW, true, true, B_FANCY_BORDER);
+	fToolBar = new ToolBar();
+	fToolBar->ChangeIconSize(16);
+	// TODO: other actions
+	fToolBar->AddAction(kMsgSort, B_TRANSLATE("Sort"), "kIconOutlineSort", true);
+	fToolBar->SetExplicitMinSize(BSize(250, B_SIZE_UNSET));
+	fToolBar->SetExplicitMaxSize(BSize(B_SIZE_UNLIMITED, B_SIZE_UNSET));
+	BLayoutBuilder::Group<>(this)
+		.AddGroup(B_VERTICAL, 0)
+			.Add(fToolBar)
+			.Add(fScrollView)
+		.End();
+}
+
+
+/* virtual */
+void
+FunctionsOutlineView::AttachedToWindow()
+{
+	BView::AttachedToWindow();
+	fToolBar->SetTarget(this);
+	if (LockLooper()) {
+		Window()->StartWatching(this, MSG_NOTIFY_EDITOR_SYMBOLS_UPDATED);
+		UnlockLooper();
+	}
+}
+
+
+/* virtual */
+void
+FunctionsOutlineView::DetachedFromWindow()
+{
+	if (LockLooper()) {
+		Window()->StopWatching(this, MSG_NOTIFY_EDITOR_SYMBOLS_UPDATED);
+		UnlockLooper();
+	}
+	BView::DetachedFromWindow();
+}
+
+
+/* virtual */
+void
+FunctionsOutlineView::Pulse()
+{
+	Editor* editor = gMainWindow->TabManager()->SelectedEditor();
+	if (editor == nullptr)
+		return;
+
+	// TODO: Not very nice: we should request symbols update only if there's a
+	// change. But we shouldn't ask one for EVERY change.
+	// Maybe editor knows, but I wasn't able to make it work there
+
+	// Update every 10 seconds
+	const bigtime_t kUpdatePeriod = 10000000LL;
+
+	// TODO: Not very nice design-wise, either:
+	// ideally we shouldn't know about Editor or LSPEditorWrapper.
+	bigtime_t currentTime = system_time();
+	if (currentTime - fSymbolsLastUpdateTime > kUpdatePeriod) {
+		editor->GetLSPEditorWrapper()->RequestDocumentSymbols();
```

Comment:
How about sending a sync message with reply?

Reply:
The retrieval of symbols could take a couple seconds, so it's a no-no.

Reply:
On a second though : yes, this symbol retrieval request could be done with a message (without reply). Lsp is asynchronous, so it would send a message back when they are ready

Reply:
Done